Enhanced Levenberg-Marquardt Algorithm with Residual Function Support

diluculo opened the pull request:
This PR implements:
- Support for direct residual functions in Levenberg-Marquardt algorithm
- Improved MINPACK-style control flow for damping parameter updates
- Comprehensive statistical metrics in `NonlinearMinimizationResult`
- Numerical differentiation through dedicated `NumericalJacobian` class

Closes #1114

diluculo committed:
Rename NonlinearObjectiveFunction to NonlinearObjectiveModel to better reflect its purpose in modeling optimization problems

diluculo committed:
Add direct residual function support to NonlinearObjectiveModel and related interfaces

diluculo committed:
Improve numerical differentiation in NonlinearObjectiveModel

- Add configurable accuracy with orders 1-6 for both model and residual functions
- Use NumericalJacobian class for more reliable derivative approximation

diluculo committed:
Add factor of 2.0 to covariance calculation to compensate for 1/2 in objective function

diluculo committed:
Apply MINPACK-style cubic damping and trust-region update to LM optimizer

diluculo committed:
Fix missing variable copies in CreateNew() method of NonlinearObjectiveModel

diluculo committed:
Add direct residual circle fitting test to NonLinearCurveFittingTests

diluculo committed:
Add comprehensive statistical metrics to NonlinearMinimizationResult

- TStatistics and PValues for parameter significance testing
- ConfidenceIntervalHalfWidths for parameter precision
- Dependencies to measure parameter correlations
- Goodness-of-fit statistics

diluculo committed:
Add ParameterStatistics class for computing parameter inference statistics

diluculo commented:
Lastly, I've added a new `ParameterStatistics` class that computes all the key statistical measures for regression parameters, including standard errors, t-statistics, p-values, confidence intervals, and dependency analysis.

Ready for review and merge.

diluculo commented:
I've just added the `ILeastSquaresMinimizer` interface, which unifies our API for nonlinear least squares minimizers by providing two overloads for the `FindMinimum` method (one for Vector<double> and one for double[]). 

Please review and merge when ready. Thank you!

diluculo commented:
In the meantime, I found a critical bug in the Trust Region optimization implementation. The subproblem solvers were using unscaled gradient and Hessian values directly from the IObjectiveModel, while they should have been using the scaled values that NonlinearMinimizerBase.EvaluateJacobian produces.

The last commit fixes this issue by modifying the ITrustRegionSubproblem interface to accept the correctly scaled gradient and Hessian as parameters.

diluculo commented:
Summary of Key Changes

1. Enhanced Levenberg-Marquardt Algorithm
   - Support for direct residual functions
   - Improved MINPACK-style control flow for damping parameter updates

1. Comprehensive Statistical Analysis
   - New `ParameterStatistics` class for regression parameters analysis
   - Calculates standard errors, t-statistics, p-values, confidence intervals, and dependencies

1. API Improvements
   - Added `ILeastSquaresMinimizer` interface to unify nonlinear least squares minimizers

1. Bug Fixes
   - Fixed a bug in Trust Region optimization where subproblem solvers were using unscaled gradient and Hessian values
   - Modified `ITrustRegionSubproblem` interface to accept correctly scaled gradient and Hessian as parameters

febkor commented:
Hi @diluculo,
Thank you for the trying to improve the optimizers here.
There's a few independent changes in this PR which makes review difficult however.

Please could this bit be pulled out to a separate PR for now?
> Fixed a bug in Trust Region optimization where subproblem solvers were using unscaled gradient and Hessian values

diluculo commented:
Hi @febkor  

Thanks for your review. You're right, this PR contains several independent changes that should be separated. I've split it into 3 separate PRs:

1. Enhanced Levenberg-Marquardt Algorithm with residual support (including ParameterStatistics)
2. ILeastSquaresMinimizer interface : #1122 
3. Trust Region optimization bug fix : #1120
4. ~~ParameterStatistics for parameter inference~~

I initially planned to separate `ParameterStatistics` as well, but since it depends on changes to `IObjectiveModelEvaluation` in the LM enhancements, I've kept it in this PR to maintain functionality.

**Edit**:
In addition to splitting the PR, I've also added new features that I had been working on:
* The Basin Hopping global optimization algorithm in PR #1122
* The Exact Trust Region subproblem solver in PR #1120

Hi @cdrnet, 
This ended up being more work than initially anticipated, but I've completed everything I had prepared. I hope these PRs can be included in v6.0.

All PRs are now ready for review.

diluculo commented:
Regarding the failing test in AppVeyor: This is unrelated to the Levenberg-Marquardt enhancements - it's a floating-point precision issue in the statistics tests for .NET 8.0 (off by ~1.03E-09).

diluculo commented:
Hi @febkor 

Here is the mathematical background document I've prepared to explain the residual function support. It illustrates the relationship between the objective function (χ²) and the residual vector (R). I hope this helps in understanding the purpose and implementation of this PR.

![image](https://github.com/user-attachments/assets/db4a849f-1d93-43b7-87fa-e7fa34f24e20)

febkor commented on src/Numerics/Optimization/NonlinearMinimizationResult.cs:
```diff
@@ -31,19 +66,69 @@ public class NonlinearMinimizationResult
             ReasonForExit = reasonForExit;
 
             EvaluateCovariance(modelInfo);
+            EvaluateGoodnessOfFit(modelInfo);
```
Both EvaluateCovariance, EvaluateGoodnessOfFit seem like they should be optional.
The user might not need these, but they are relatively expensive to calculate.
On the master branch, EvaluateCovariance for example shows up in profiles as being quite significant (esp due to its allocations).

febkor commented on src/Numerics/Optimization/LevenbergMarquardtMinimizer.cs:
```diff
@@ -140,91 +141,181 @@ public NonlinearMinimizationResult Minimum(IObjectiveModel objective, Vector<dou
                 while (true)
                 {
+                    // Store current Hessian diagonal for restoration if step is rejected
+                    var savedDiagonal = Hessian.Diagonal().Clone();
```
Is the separate `Clone()` required?
Matrix.Diagonal() produces a new vector with values copied in from the matrix.

febkor commented on src/Numerics/Optimization/LevenbergMarquardtMinimizer.cs:
```diff
@@ -140,91 +141,181 @@ public NonlinearMinimizationResult Minimum(IObjectiveModel objective, Vector<dou
+            var mu = initialMu * Hessian.Diagonal().Max(); // Damping parameter μ
+            var nu = 2.0; // Multiplication factor ν for mu updates
+
+            // Counters for successful and failed iterations
```
The `ncsuc` counter seems to get incremented, but is not used for any logic?

febkor commented on src/Numerics/Optimization/ObjectiveFunctions/NonlinearObjectiveModel.cs:
```diff
@@ -0,0 +1,688 @@
+
+                // Create adapter function that returns the model value for current observation
+                // when given parameters array
+                double funcAdapter(double[] p)
```
It seems it'd be more natural to extend the NumericalJacobian to support modelFunctions of the shape needed here?
The usage here seems more complex than needed and some inefficiencies:
- _modelFunction returns a whole vector, but only a single value is used
- this funcAdapter lambda creates an allocation per observation (minor in the context of the rest of the code, but still avoidable)

febkor commented on src/Numerics/Optimization/ObjectiveFunctions/NonlinearObjectiveModel.cs:
```diff
@@ -0,0 +1,688 @@
+                weights = weights.PointwiseAbs();
+            }
+
+            Weights = (weights == null)
```
Does the `Weights` matrix get used anywhere?
Apart from some trivial null checks, the usages I noticed all only use the diagonal, i.e. the `weights` vector. Are there any I might have missed?
It can be quite expensive in memory terms to create it.

febkor reviewed (commented):
Just a brief review, more from a performance angle.
Hope you find them useful.
